coderd/workspaceagentsrpc.go

```diff
@@ -261,84 +307,77 @@ func (api *API) agentConnectionUpdate(ctx context.Context, workspaceAgent databa
 		defer cancel()
 
 		// Only update timestamp if the disconnect is new.
-		if !disconnectedAt.Valid {
-			disconnectedAt = sql.NullTime{
+		if !m.disconnectedAt.Valid {
+			m.disconnectedAt = sql.NullTime{
 				Time:  dbtime.Now(),
 				Valid: true,
 			}
 		}
-		err := updateConnectionTimes(ctx)
+		err := m.updateConnectionTimes(finalCtx)
 		if err != nil {
 			// This is a bug with unit tests that cancel the app context and
 			// cause this error log to be generated. We should fix the unit tests
 			// as this is a valid log.
 			//
 			// The pq error occurs when the server is shutting down.
 			if !xerrors.Is(err, context.Canceled) && !database.IsQueryCanceledError(err) {
-				api.Logger.Error(ctx, "failed to update agent disconnect time",
+				m.logger.Error(finalCtx, "failed to update agent disconnect time",
 					slog.Error(err),
-					slog.F("workspace_id", workspaceID),
 				)
 			}
 		}
-		api.publishWorkspaceUpdate(ctx, workspaceID)
+		m.updater.publishWorkspaceUpdate(finalCtx, m.workspaceBuild.WorkspaceID)
+	}()
+	reason := "disconnect"
+	defer func() {
+		_ = m.conn.Close(websocket.StatusGoingAway, reason)
 	}()
 
-	err := updateConnectionTimes(ctx)
+	err := m.updateConnectionTimes(ctx)
 	if err != nil {
-		_ = conn.Close(websocket.StatusGoingAway, err.Error())
-		return nil, false
+		reason = err.Error()
+		return
 	}
-	api.publishWorkspaceUpdate(ctx, workspaceID)
-
-	return func(closeCtx context.Context, ensureLatestBuildFn func() error) {
-		ticker := time.NewTicker(api.AgentConnectionUpdateFrequency)
-		defer ticker.Stop()
-		for {
-			select {
-			case <-closeCtx.Done():
-				return
-			case <-ticker.C:
-			}
+	m.updater.publishWorkspaceUpdate(ctx, m.workspaceBuild.WorkspaceID)
+
+	ticker := time.NewTicker(m.pingPeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			reason = "canceled"
+			return
+		case <-ticker.C:
+		}
 
-			lastPing := *lastPing.Load()
-
-			var connectionStatusChanged bool
-			if time.Since(lastPing) > api.AgentInactiveDisconnectTimeout {
-				if !disconnectedAt.Valid {
-					connectionStatusChanged = true
-					disconnectedAt = sql.NullTime{
-						Time:  dbtime.Now(),
-						Valid: true,
-					}
-				}
-			} else {
-				connectionStatusChanged = disconnectedAt.Valid
-				// TODO(mafredri): Should we update it here or allow lastConnectedAt to shadow it?
-				disconnectedAt = sql.NullTime{}
-				lastConnectedAt = sql.NullTime{
-					Time:  dbtime.Now(),
-					Valid: true,
-				}
-			}
-			err = updateConnectionTimes(ctx)
-			if err != nil {
-				_ = conn.Close(websocket.StatusGoingAway, err.Error())
-				return
-			}
-			if connectionStatusChanged {
-				api.publishWorkspaceUpdate(ctx, workspaceID)
-			}
-			err := ensureLatestBuildFn()
-			if err != nil {
-				// Disconnect agents that are no longer valid.
-				_ = conn.Close(websocket.StatusGoingAway, "")
-				return
-			}
+		lastPing := *m.lastPing.Load()
+		if time.Since(lastPing) > m.disconnectTimeout {
```

Comment:
For testing, a customizeable now function could be nice.

Reply:
We could introduce mock points for all the time related stuff, control the ticks, and the times returned, etc.  But, that's a bit of overhead, and it means that the tests don't call the same code as prod.  For this case, I think it's more understandable in testing to just have customizable periods and use the "real" `time` functions.